odoo/tools/misc.py

```diff
@@ -1215,9 +1215,14 @@ def get_lang(env, lang_code=False):
     :return res.lang: the first lang found that is installed on the system.
     """
     langs = [code for code, _ in env['res.lang'].get_installed()]
-    for code in [lang_code, env.context.get('lang'), env.user.company_id.partner_id.lang, langs[0]]:
-        if code in langs:
-            return env['res.lang']._lang_get(code)
+    lang = langs[0]
```

Comment:
or use generator ?

Reply:
not sure how to apply that on an array loop

Reply:
with a func that yield 1, yield 2, yield 3...
But I prefere you if solution than a fake function ;)

Reply:
+1 for `if`s instead of a generator. Simpler, more readable, faster (no coroutine to setup, etc.)

While you're unrolling the loop, you could micro-benchmark having an extra set for the `in` tests (as it is O(n) on lists). Most databases will have only 1 lang installed, so the cost of creating a one-element set should not offset the benefit of using it, but some databases have 16 languages installed, so it's worth a quick comparison ;-)

Also the first `in` test can be short-circuited when `lang_code` is not provided, which is the default.

Reply:
I just did a micro benchmark as you suggested with the following diff:

```diff
--- odoo/tools/misc.py
+++ odoo/tools/misc.py
@@ -1216,7 +1216,8 @@ def get_lang(env, lang_code=False):
     """
     langs = [code for code, _ in env['res.lang'].get_installed()]
     lang = langs[0]
-    if lang_code in langs:
+    langs = set(langs)
+    if lang_code and lang_code in langs:
         lang = lang_code
```

and run `get_lang(env)` **10000x** with and without the `langs = set(langs)`, with 1 and 16 languages.

And change doesn't seems impact performance

Result:

> With set & 1 lang: 0.09833766400697641
> Without set & 1 lang: 0.09716381199541502
> 
> With set & 16 lang: 0.1100444290204905
> Without set & 16 lang: 0.10154330500517972
> 

So I suppose that we can add or not the set without any impact. What you will prefer will be what I will do, but since there are no change, do we really need to make an extra op ;) PErso I will let it like that.


And yes, we were all agrees about the generator. It was my first idea when I see the 'imp', but we fastly mark as resolved this comment since it was no more elegant, no more powerful to use it. And that we prefer explicit ifs in this case. So we are on the same 'wave length' :) :+1:

Reply:
Since the place of the 'matched lang' could have an impact in the list, I suppose that if first or last in the list, the matching could be faster.
Just re-tested now with 30 langs, with the 'matched lang' in the middle of this list and always no impact.